nvidia-update.sh

```diff
@@ -28,7 +28,7 @@ function usage() {
 }
 
 function mktemppkg() {
-	echo "$(mktemp $TMPDIR/$(uuidgen).pkg)"
+	echo "$(/usr/bin/mktemp $TMPDIR/$(uuidgen).pkg)"
```

Comment:
Now don't get me more, I'm not sure why a user would override `mktemp`, but who am I to ignore a users setup?

Reply:
I am exactly a user than override the default names from coreutils. Like you said there, if you point to those provided by macOS then it's guaranteed to works.

Reply:
The _script_ is guaranteed to work (have valid syntax in system bash), which is important.

However, if a user is replacing coreutils they should _not_ be violating the contract, however users are free to alter the functionality.

In this case a user might want their temp directory to run on a different volume etc. I don't like violating a user's preferences. This is different from the script itself using the system bash, as a working shell is required for the script to run at all (syntax must be accepted by the shell), however I see no reason to disallow user customisation if that's what the user wishes.

Reply:
I mean, I got error if I don't use the system's mktemp. The reason is as the GNU tools are updated, backward-incompatible changes are made. I can give you the error msg. if you want.

Alternatively, the end-user can override their PATH to use the system's default instead. But since you are sure the tools comes with macOS works, pointing to them directly are easier. And in this case I only changed those that an error is found explicitly. i.e. I'm not brute-force specifying any executables with absolute paths.

Reply:
Ok, this is the error when the mktemp from coreutils is used:

```bash
mktemp: too few X's in template ‘/var/folders/s6/z979kbkn5dn_2ggfywznw9yw0000gn/T//96BB5DC8-B013-4C6C-BAB4-97A4CC3C9D0E.pkg’
```

Reply:
Hmm, I'd be inclined to say this is an error on your system as you've violated the contract of macOS' `mktemp` by using a version that requires templates to end in `XXX`.

However, I must admit the use of `mktemp` at all in this circumstance is rather dubious. We've given an exact file path, and are not utilising the template behaviour of `mktemp` at all. I opted to use `mktemp` for semantic reasons, but perhaps that was ill advised.

I think simply using `touch` may be a better solution.

Reply:
> Hmm, I'd be inclined to say this is an error on your system as you've violated the contract of macOS' mktemp by using a version that requires templates to end in XXX.

You just said it is up to the user to customized their environment. Since macOS is frozen at GPLv2, replacing them with up-to-date versions in GPLv3 is like a decade in advance and certainly something will break. The user customizes their environment for their own use. But a script like this should relies on stable binaries provided by the system instead. In macOS's case it is actually easy, since if you point it to the abs. path in `/usr/bin`, etc. then they are almost guaranteed to behave uniformly.

But it is up to you if you want to merge this (I reduced this PR this that single commit.) I already have ways to deal with it personally. I just want to make sure others won't encounter this problem.

Reply:
> You just said it is up to the user to customized their environment. Since macOS is frozen at GPLv2, replacing them with up-to-date versions in GPLv3 is like a decade in advance and certainly something will break. The user customizes their environment for their own use. But a script like this should relies on stable binaries provided by the system instead. In macOS's case it is actually easy, since if you point it to the abs. path in /usr/bin, etc. then they are almost guaranteed to behave uniformly.

Updating/customising your system is one thing, breaking contracts is _another thing entirely_.

There's absolutely nothing stopping you customising your system's behaviour if you do so in a way that violates no contracts i.e. inputs and outputs should match their documented (and sometimes nuanced) forms. However, you're free to modify the encapsulated functionality to your heart's content. In a software sense, your changes _must_ be backwards compatible - you can add new functionality, alter the behaviour of existing functionality; but _not alter the interface of existing functionality_.

You can't reasonably expect your OS or third-party software (such as this script) to continue functioning normally when you're violating software contracts in a system-wide fashion. It's certainly okay to do so in a contained environments e.g. a dedicated shell for compiling GNU software. However, doing so system-wide is quite simply akin to _breaking your system_.

I'm comfortable being compatible with a system that is itself backwards compatible, however I'm not too interested in trying to be forward compatible with every possible system variation. If users are customising their system, there's nothing stopping them disabling SIP and replacing system binaries anyway. The line has to be drawn somewhere.

Reply:
The situation is not as simple as you think. There are plenty of bash scripts out there, and many of them written to works on Linux, and some of them written to work across different OSes (though for them), and some of them written only for macOS for various reasons (for example they don't use OS other than macOS, or the target platform is mac only like this one.) So by upgrading the default tools to the latest versions, it should means that the first 2 cases are pretty much covered. For the last case, since they target mac only, their are obvious ways to make it work even with such customization. e.g. `/bin/sh` is bash uses such knowledge.

I don't understand why a change like this that will not break anyone's installation for those whose system the script would already work, but add a safety for some cornering cases will not be accepted. Anyway, the repo is yours, good day.